unique_toolkit/unique_toolkit/evaluators/hallucination/utils.py

```diff
@@ -0,0 +1,201 @@
+        logger (Optional[logging.Logger], optional): The logger used for logging information and errors. Defaults to the logger for the current module.
+
+    Returns:
+        EvaluationMetricResult | None: The result of the evaluation, indicating the level of hallucination. Returns `None` if the metric is not enabled or if an error occurs.
+
+    Raises:
+        EvaluatorException: If the context texts are empty, required fields are missing, or an error occurs during the evaluation.
+    """
+    model_name = config.language_model.name
+    logger.info(f"Analyzing level of hallucination with {model_name}.")
+
+    input.validate_required_fields(hallucination_required_input_fields)
+
+    try:
+        msgs = _get_msgs(input, config, logger)
+        result = await LanguageModelService.complete_async_util(
+            company_id=company_id, messages=msgs, model_name=model_name
+        )
+        result_content = result.choices[0].message.content
+        if not result_content:
+            error_message = "Hallucination evaluation did not return a result."
+            raise EvaluatorException(
+                error_message=error_message,
+                user_message=error_message,
+            )
+        return parse_eval_metric_result(
+            result_content, EvaluationMetricName.HALLUCINATION
+        )
+    except Exception as e:
+        error_message = "Error occurred during hallucination metric analysis"
+        raise EvaluatorException(
+            error_message=f"{error_message}: {e}",
+            user_message=error_message,
+            exception=e,
+        )
+
+
+def _get_msgs(
+    input: EvaluationMetricInput,
+    config: EvaluationMetricConfig,
+    logger: logging.Logger,
+):
+    """
+    Composes the messages for hallucination analysis based on the provided input and configuration.
+
+    This method decides how to compose the messages based on the availability of context texts and history
+    message texts in the `input`
+
+    Args:
+        input (EvaluationMetricInput): The input data that includes context texts and history message texts
+                                      for the analysis.
+        config (EvaluationMetricConfig): The configuration settings for composing messages.
+        logger (Optional[logging.Logger], optional): The logger used for logging debug information.
+                                                     Defaults to the logger for the current module.
+
+    Returns:
+        The composed messages as per the provided input and configuration. The exact type and structure
+        depend on the implementation of the `compose_msgs` and `compose_msgs_default` methods.
+
+    """
+    if input.context_texts or input.history_messages:
+        logger.debug("Using context / history for hallucination evaluation.")
+        return _compose_msgs(input, config)
+    else:
```

Comment:
This else might be superfluous, since there is already an early return in the `if` block.